guides/internals.rst

```diff
@@ -0,0 +1,342 @@
+===================================
+How to Work on Hypothesis Internals
+===================================
+
+This is a guide to how to work on Hypothesis internals,
+with a particular focus on helping people who are new to it.
+Right now it is very rudimentary and is intended primarily for people who are
+looking to get started writing shrink passes as part of our `current outreach
+program to get more people doing that <https://github.com/HypothesisWorks/hypothesis-python/issues/1093>`_,
+but it will expand over time.
+
+------------------------
+Bird's Eye View Concepts
+------------------------
+
+The core engine of Hypothesis is called Conjecture.
+
+The "fundamental idea" of Conjecture is that you can represent an arbitrary
+randomized test case as a string of bytes, which are basically intended as the
+underlying entropy of some pseudo-random number generator (PRNG).
+Whenever you want to do something "random" you read the next bytes and
+do what they tell you to do. By manipulating these bytes, we can achieve
+more interesting effects than pure randomness would allow us to do, while
+retaining the power and ease of use of random testing.
+
+The idea of shrinking in particular is that once we have this representation,
+we can shrink arbitrary test cases based on it. We try to produce a string that
+is *shortlex minimal*. What this means is that it has the shortest possible
+length and among those strings of minimal length is lexicographically (i.e. the
+normal order on strings - find the first byte at which they differ and use that
```

Comment:
typographical nitpicks, em-dash (—) for punctuation dash, also on lines188, 338, comma after i.e.

Reply:
Due to annoying technical constraints (because everything is awful despite it being 2018 my console is not unicode-sensible, I'm unwilling to write this as `&mdash;` and Github doesn't render RST em-dashes correctly) they're not going to be  em-dashes. 😢 

Elaborate on comma after i.e.? Is that a thing I'm supposed to bee doing?

Reply:
Could we all chip in and buy you a proper terminal? :smile:  Re i.e., I had that beaten into me from childhood, but a quick snout around reveals it is mandatory for US English (if the Chicago manual of style is your Bible, it is mine), but optional in BE (Fowler); so I have been deceived by yankee cultural imperialism.  I'll continue to use though, since the comma indicates the natural spoken pause that you would use when replacing it with "that is".

Reply:
> Could we all chip in and buy you a proper terminal? 

Unfortunately due to a complex interlocking series of least worst solutions and bad life choices, getting me a better terminal would require getting me a better computer (I am currently working inside the Windows Ubuntu subsystem, and all of the terminal options for Windows are awful).